lib/normalizer.c

```diff
@@ -1967,6 +1967,53 @@ grn_nfkc_normalize_unify_hiragana_voiced_sound_mark(
   return utf8_char;
 }
 
+static inline const unsigned char *
+grn_nfkc_normalize_add_hiragana_voiced_sound_mark(
+  const unsigned char *utf8_char, unsigned char *voiced)
+{
+  if (utf8_char[0] == 0xe3) {
+    if ((utf8_char[1] == 0x81 &&
+         (0x8b <= utf8_char[2] && utf8_char[2] <= 0xa1))) {
+      /* U+304B HIRAGANA LETTER KA ..
+       * U+3061 HIRAGANA LETTER TI */
+      if (utf8_char[2] & 0x1) {
+        voiced[0] = utf8_char[0];
+        voiced[1] = utf8_char[1];
+        voiced[2] = utf8_char[2] + 1;
+        return voiced;
+      }
+    } else if ((utf8_char[1] == 0x81 &&
+                (0xa4 <= utf8_char[2] && utf8_char[2] <= 0xa8))) {
+      /* U+3064 HIRAGANA LETTER TU ..
+       * U+3068 HIRAGANA LETTER TO */
+      if (!(utf8_char[2] & 0x1)) {
+        voiced[0] = utf8_char[0];
+        voiced[1] = utf8_char[1];
+        voiced[2] = utf8_char[2] + 1;
+        return voiced;
+      }
+    } else if ((utf8_char[1] == 0x81 &&
+                (0xaf <= utf8_char[2] && utf8_char[2] <= 0xbd))) {
+      /* U+306F HIRAGANA LETTER HA ..
+       * U+307D HIRAGANA LETTER PO */
+      unsigned char mod3 = (unsigned char)((utf8_char[2] - 0xaf) % 3);
+      if (mod3 == 0) {
+        voiced[0] = utf8_char[0];
+        voiced[1] = utf8_char[1];
+        voiced[2] = utf8_char[2] + 1;
+        return voiced;
+      } else if (mod3 == 2) {
+        voiced[0] = utf8_char[0];
+        voiced[1] = utf8_char[1];
+        voiced[2] = utf8_char[2] - 1;
+        return voiced;
+      }
+    }
+  }
```

Comment:
Does the added tests cover all of these cases?

FYI: The `unify_kana_voiced_sound_mark` test: https://github.com/groonga/groonga/blob/main/test/command/suite/normalizers/nfkc/unify_kana_voiced_sound_mark_hiragana.test

Reply:
fix: [27b6ce9](https://github.com/groonga/groonga/pull/2515/commits/27b6ce93357cc14a50e499d1543828092d3037a6) I added the test cases cover all of above cases and boarders.